src/triage/util/db.py

```diff
@@ -0,0 +1,19 @@
+from sqlalchemy import create_engine
+
+
+class SerializableDbEngine(object):
+    def __init__(self, db_engine):
+        self._db_engine = db_engine
+        if db_engine:
+            self._db_engine_url = db_engine.url
+
+    @property
+    def db_engine(self):
+        if not self._db_engine and self._db_engine_url:
+            self._db_engine = create_engine(self._db_engine_url)
+        return self._db_engine
+
+    def __getstate__(self):
+        if self._db_engine:
+            self._db_engine = None
+        return self.__dict__.copy()
```

Comment:
That said, I see that `create_engine` [[docs]](http://docs.sqlalchemy.org/en/latest/core/engines.html#sqlalchemy.create_engine) accepts more arguments than merely the connection URI.

I'm concerned that, as is, this won't preserve all configuration options the user may have set on their engine.

Moreover, if we don't and won't support provision of anything more than a connection URI, then why do we insist that they provide an engine object, (rather than just the URI string)? If the user could only provide a connection string, then we could, if we liked, go even further, _e.g._ only storing that connection string on the prediction object, and creating the engine on every property access, rather than extending/wrapping the engine class, (though I don't think we need to go that far, since there are a lot of options).

Rather, it seems useful to me that we allow them to provide any compatible engine object they like, but that we not assume its pickle state to be just its connection URI. Maybe you see how to extrapolate other arguments, though I'm not sure, myself.

Alternatively, we could let the user supply an engine-creating function, which would mean we don't need to pickle anything, or we can make pickling very simple.

Or we could offer our own `create_engine`, which handles it all:

    class SerializableDbEngine(wrapt.ObjectProxy):

        __slots__ = ('url', 'creator', 'kwargs')

        def __init__(self, url, *, creator=sqlalchemy.create_engine, **kwargs):
            self.__setstate__((url, creator, kwargs))

        def __getstate__(self):
            return [object.__getattribute__(self, key) for key in self.__slots__]

        def __setstate__(self, state):
            for (key, value) in zip(self.__slots__, state):
                object.__setattr__(self, key, value)

            (url, creator, kwargs) = state
            engine = creator(url, **kwargs)
            super().__init__(engine)

    create_engine = SerializableDbEngine

…and users may then use our `triage.util.db.create_engine` just as they would otherwise use `sqlalchemy.create_engine`, (though in that case we might want to either house this or at least import it under a higher-level namespace, such as `triage.create_engine`).

Reply:
You're right that this is a problem. I did try to find some way to infer more about the instantiated engine object so that information could be serialized too, but the engine's `__dict__` didn't have much. If at all possible I don't want to force people to pass in something (a connection url, a callable, a triage namespaced engine object) other than a plain old sqlalchemy engine: it's what other tools use, and it's one thing I'd like to keep standard. However, I don't know if this is a real option; in that case, our own create_engine as you defined above is probably the best option.

Reply:
How about this –

    class SerializableDbEngine(wrapt.ObjectProxy):
    
        __slots__ = ('url', 'creator', 'kwargs')
    
        def __init__(self, url, *, creator=sqlalchemy.create_engine, **kwargs):
            self.url = url
            self.creator = creator
            self.kwargs = kwargs
    
            engine = creator(url, **kwargs)
            super().__init__(engine)
    
        def __reduce__(self):
            return (self.__reconstruct__, (self.url, self.creator, self.kwargs))
    
        @classmethod
        def __reconstruct__(cls, url, creator, kwargs):
            return cls(url, creator=creator, **kwargs)

    create_engine = SerializableDbEngine

Works for me.